\documentclass[11pt]{article}

\usepackage{fullpage}
\usepackage[round,semicolon]{natbib}
\usepackage{amssymb}
\usepackage{amsmath}
\usepackage{bbm}

\newcommand{\pr}{\text{Pr}}

\newcommand{\ETAL}{\textit{et al.}}
\newcommand{\EG}{\textit{e.g.}}
\newcommand{\IE}{\textit{i.e.}}
\newcommand{\CF}{\textit{c.f.}}

\usepackage{hyperref}

% NOTE: one sentence per line for nice git diffs
% WD: I'm a fan of source comments like this, instead of the colory ones in the pdf. But Ok fine, have it your way!
\usepackage{color}
% Will comments
\newcommand{\wdcomment}[1]{{\color{magenta}{(\textbf{WD's comment:} #1)}}}
% Andy comments
\newcommand{\amcomment}[1]{{\color{blue}{(\textbf{AM's comment:} #1)}}}
% Sarah comments
\newcommand{\shcomment}[1]{{\color{red}{(\textbf{SH's comment:} #1)}}}


\title{Phyload: a story of paired sites finding their way in an unpaired model space}

% WD: this alphabetical author list is copied from the previous version, not meant to indicate priority
\author{
William DeWitt$^{1\ast}$, Sarah Hilton$^{1\ast}$, and Andrew Magee$^{2\ast}$\\
\small{Departments of $^1$Genome Sciences and $^2$Biology, University of Washington, Seattle, USA}\\
\small{$^\ast$ Equal contribution}
}
% \author[1]{Sarah Hilton}
% \author[2]{Andy Magee}

% \affil[1]{Department of Genome Sciences, University of Washington, Seattle, USA}
% \affil[2]{Department of Biology, University of Washington, Seattle, USA}

\begin{document}

\maketitle

\amcomment{We should take the time to consider clearing up our nomenclature. Sites, site pairs, epistatic sites, iid sites, all of these are things that could easily get us into trouble when writing/communicating our results to people.}

\begin{abstract}
Phylogenetic inference relies on probabilistic models of character state change along tree branches in which sites in the alignment are statistically independent.
This severely restrictive assumption facilitates computational tractability, but cannot accommodate epistatic coupling of sites during the evolution of functional sequences that arise from structural and functional constraints on the encoded molecule.
We consider the effect of this misspecification error on the accuracy of phylogenetic reconstruction in a setting of pairwise epistasis.
We show that including epistatically coupled sites in an alignment improves reconstruction accuracy and we introduce an alignment test statistic that is diagnostic for epistasis and can be used in poster predictive checks.
\end{abstract}

\section*{{Introduction}\label{sec:intro}}

Epistasis is $<$definition$>$
This phenomenon is pervasive in datasets for phylogenetic inference $<$citations$>$

Phylogenetic models of epistasis have a long history... we should be sure to emphasize here that most of these are pair models.
$<$We need to talk up the NH model here a decent bit$>$
\amcomment{We also need to be clear if we're going to discuss it as a model that has both iid and interacting sites or not, so we can be clear about our terminology when we set up our grid}

However, phylogenetic models of epistasis are not without drawbacks.
When applying these models in practice, most require that a user specify \textit{a priori} all pairs of sites that are interacting $<$cite doublet models, NH model, pre-RJ Salamin models, others if we come across them$>$.
In some cases such as identifying pairs of stem sites in ribosomal RNA (rRNA), this is possible, if difficult $<$citations on the shitshow of annotating things? maybe some of the databases?$>$.
Even these models of epistasis are more computationally demanding than models that assume all sites are iid by increasing the size of the rate matrix from $4 \times 4$ to $16 \times 16$ and thus
Nonetheless, in many cases there is no hope for pre-defining pairs of interacting sites, which lead \cite{meyer2019simultaneous} to develop a suitable extension of epistatic models for the case where interacting sites must be inferred.
The cost of their approach is that the dimensionality of the parameter space is both drastically increased and variable, requiring reversible-jump MCMC \citep{green1995reversible} to sample from the posterior distribution of the assignments of sites to interacting pairs.

In parallel to the literature on phylogenetic models of epistasis, a literature on detecting epistasis in phylogenetic datasets.
$<$Citations and some literature reviewing$>$
Something something but this literature has mostly been confined to detecting the presence of epistatic interactions in datasets, and decoupled from either modeling these interactions or asking whether they exert a notable effect on phylogenetic inference.
Mumble mumble they should have been doing posterior-predictive testing, or at least using alignment-based summary statistics because those are better because reasons.

Given the difficulties involved in applying epistatic phylogenetic models to real datasets, whether broadly applicable \citep{meyer2019simultaneous} or more specific to the dataset at hand $<$doublet and NH citations$>$, we seek to answer three questions involving the use of misspecified (site-iid) models for dataset where epistasis exists.
First and foremost, what is the effect of including epistatically paired sites in inference?
Secondly, can we detect the presence of epistatic interactions in datasets using posterior-predictive simulations?
Thirdly, is there a test-statistic or set of test-statistics that allow researchers to detect the presence of specifically problematic epistasis?
That is, if there is epistasis in a dataset and failing to model is is causing poor or inadequate phylogenetic inference, is there a way to know we are in this regime?

\amcomment{okay but can we rename this, d is such a poor choice of letter and is so confusing to talk about}
To address these questions, we perform a simulation study using the epistatic model of \cite{nasrallah2013phylogenetic}, which has a single parameter, $d$ that controls the strength of epistatic interactions.
We simulate alignments on a 3-dimensional grid, defined by the number of sites in the alignment that site-iid, the number of sites(not site-pairs) that are drawn from a paired model, and the value of $d$.
Our grid thus includes a variety of alignment sizes.
This allows us to answer our first question and determine whether the inclusion of epistatic sites in an alignment improves or worsens phylogenetic estimates by tracing a line (for a fixed value of $d$) of increasing numbers of epistatic sites for a fixed value of iid sites.
We define a set of test statistics, including several new ones designed to pick up on patterns of epistasis, and several previously described more general test statistics.
By comparing across levels of epistasis in our simulated alignments, and by their use in posterior predictive checks, we address whether epistasis can be detected directly from alignments, answering our second question.
By combining the posterior-predictive analyses used to answer question 2 and the accuracy results used to answer question 1, we can address whether we can detect problematic epistasis, answering our third question.


Look at all the dumb shit physics/math/CS people have been saying about inferring phylogenies being impossible because selection.
Will's pal (who will remain unnamed) deigned to opine, but upon further discussion revealed that he doesn't know what a codon is, or a synonymous mutation.
We will crush them.

It would be nice to write an intro to phylo modeling that builds up complexity, mentioning GTR+$\Gamma$ and ExpCMs, then the \cite{nasrallah2013phylogenetic} stuff.

\section*{Model\label{sec:Model}}

\subsection*{Goal and Housekeeping}

Andy re-implemented the model from \cite{nasrallah2013phylogenetic}, which models pair-wise epistasis due to RNA secondary structure.
Below is the model as Andy implemented it in \texttt{RevBayes}.

\subsection*{RNA epistasis model}

Here is the model from \cite{nasrallah2013phylogenetic}.
We should change this around in an order we like and with notation we like.

$\boldsymbol{Q}$ is the instantaneous rate matrix describing changes from doublet $\boldsymbol{x}$ to doublet $\boldsymbol{y}$.
For $\boldsymbol{x} = (x_1, x_2)$, $x_1$ is the 5' nucleotide and $x_2$ is the 3' nucleotide.
\begin{equation}
\label{eq:Q}
\boldsymbol{Q} = \xi \times
\begin{cases}
   \pi_{\boldsymbol{y}} S_{x_1, y_1} & \mbox{if single substitution at the 5' site,} \\
   \pi_{\boldsymbol{y}} S_{x_2, y_2} & \mbox{if single substitution at the 3' site,} \\
   \pi_{\boldsymbol{y}} S_{x_1, y_1} S_{x_2, y_2} d & \mbox{if double substitution where ${\boldsymbol{x}}$ and ${\boldsymbol{y}}$ $\in {\boldsymbol{W}}$,} \\
   0 & \mbox{if any other double substitution,} \\
   - \sum_{\boldsymbol{x} \ne \boldsymbol{y}} Q_{\boldsymbol{x},\boldsymbol{y}}& \mbox{if $\boldsymbol{x}$ = $\boldsymbol{y}$} \\
   \end{cases}
\end{equation}
$\boldsymbol{S}$ is the GTR exchangeability matrix \citep{tavare1986some} and $S_{x_i,y_i}$ is understood to be the element in $S$ governing the rate of exchangeability between nucleotide $x_i$ and $y_i$ (by definition, $S_{x_i,y_i} = S_{y_i,x_i}$),
${\boldsymbol{W}} = {AT, CG, GC, TA}$ is the set of Watson-Crick pairs,
$\boldsymbol{\pi} = (\pi_{AA}, \pi_{AC}, ..., \pi_{TT})$ are the stationary state frequencies of the 16 possible doublet states,
$d$ controls rate of double to single mutations between doublets,
and $\xi$ is the rate-scaling factor.

\paragraph{Points to clarify}
\begin{itemize}
\item How do we interpret $d$? $d$ is the relative rate of double to single mutations between doublets or the ``strength" of epistatic interactions.

Nasrallah says relative proportion, we've been saying relative rate.
Are these the same?

Here is a copy of table 1 from \cite{nasrallah2013phylogenetic}:

\begin{tabular}{ |p{1cm}|p{5cm}|p{5cm}|  }
 \hline
& $\pi_{\boldsymbol{y}} = \pi_{y_1}\pi_{y_2}$&$\pi_{\boldsymbol{y}} \ne \pi_{y_1}\pi_{y_2}$\\
 \hline
$d=0$  & Independent and nonepistatic & Model inadequacy \\
\hline
 $d > 0$  & Dependent but nonepistatic & Dependent and epistatic\\
 \hline
\end{tabular}

\item How do we normalize $\boldsymbol{Q}$ ($\xi$)?

From slack: 0.5 * Pr(single) + Pr(double)

It is not completely clear how $\xi$ is defined in the paper.
Assuming it's defined in the usual way (which RevBayes can do for us), simply take a weighted sum of the off-diagonal elements $\xi^{-1} = \sum_{\boldsymbol{x}} \sum_{\boldsymbol{y} \ne \boldsymbol{x}} \pi_{\boldsymbol{x}} Q_{\boldsymbol{x},\boldsymbol{y}}$.
However, this normalizes the rate matrix on paired sites, which would count both doublet substitutions and single-substitutions equally.
This formulation does not guarantee that the number of expected substitutions per unpaired epistatic site is the same as per non-epistatic site.
Thus it seems that the appropriate normalization should be defined by,
\begin{align*}
\xi^{-1} = \sum_{\boldsymbol{x}} \sum_{\boldsymbol{y} \ne \boldsymbol{x}} \pi_{\boldsymbol{x}} \times
\begin{cases}
   \frac{1}{2}Q_{\boldsymbol{x},\boldsymbol{y}} & \mbox{if single substitution,} \\
   Q_{\boldsymbol{x},\boldsymbol{y}} & \mbox{if double substitution of the allowed type,} \\
   \end{cases}
\end{align*}
This \textit{should} recognize the fact that single substitutions change only one site in the pair.
We have to do this normalization ourselves, and can't just make RevBayes do it for us, but that's more me whining than anything important.

Looking at the NH paper, the $\xi$ ``scaling factor'' is a unified rate parameter for both $Q$ and $Q^*$, but for $Q$ the worry is that it's controlling expected numbers of events on pairs of sites, not on individual sites like for $Q^*$.
To sort out what's up with $\xi$ it we'd want to compute the expected number of single subs per site in a pair as
\[
\mathbbm{E}_\mathbf{Q}[\text{number of pair events}]\left(\frac{1}{2}\mathbbm{P}(\text{single sub})+\mathbbm{P}(\text{double sub})\right),
\]
then demand that this equals the expected number of subs per site for the null model
\[
\mathbbm{E}_\mathbf{Q^*}[\text{number of events}].
\]


\item Given the fit parameter values for a GTR model, how do we simulate under this model?

From github, ``
This is the tree inferred under GTR+GAMMA using RAxML version 8.2.12.
Inferred model parameters

alpha shape parameter = 0.440894

relative exchange rates (ac ag at cg ct gt) = 1.882161 7.009179 0.914813 0.495852 7.666181 1.000000

base frequencies = 0.340152 0.190828 0.225045 0.243974"

Under the parameterization in the paper (standard MrBayes/RevBayes parameterization), we first take the RER and simplex them, yielding $\boldsymbol{r} = (0.0992,0.3695,0.0482,0.0261,0.4042,0.0527)$.
Then we make the symmetric GTR rate matrix $\boldsymbol{S}$ from the relative rates (for entirely too much detail, we put the elements of $\boldsymbol{r}$ into the upper diagonal of $\boldsymbol{S}$ row-wise, and the lower-diagonal column-wise, to make the symmetric exchangeability matrix).
Then we draw the doublet stationary frequencies $\boldsymbol{\pi}$ from a Dirichlet(2,...,2) distribution.
Then we assemble an unscaled version of 16 x 16 matrix $\boldsymbol{Q}$ as described \ref{eq:Q}, while calculating $\xi$.
Then we normalize $\boldsymbol{Q}$, tell RevBayes that's our rate matrix on the flu tree, and draw from the induced phylogenetic CTMC distribution.
Rev simulates these as characters 0,1,2,...,D,E,F.
We also simulate the non-epistatic sites, and with an R script we turn the 16-state characters into pairs of sites, and add them to the nonepistatic sites.

\end{itemize}


\section*{Methods}
We sconsed the shit out of our simulation study. Also we used RevBayes for things.

\subsection*{Posterior predictive assessments}

One of our key questions is, can the presence of unmodeled epistatic interactions be detected with posterior predictive checks?
$<$ Explain posterior predictive stuff here $>$

\amcomment{if we can't get this alliteration in the final paper I'll be very sad}
As epistasis has yet to be studied from a posterior predictive perspective, there are currently no test posterior preictive approaches designed explicitly to detect it.
In the following sections, we describe two test statistics designed to capture features of epistatis, and a frequentist test for which we calculate the posterior distribution on p-values.
These measures are all inspired by the idea of compensatory substitutions: in a pairwise epistatic interaction, when one site is changed, the other site will most likely change as well.
This idea underlies the parameter $d$ in both the model of \cite{nasrallah2013phylogenetic} and the model of \cite{meyer2019simultaneous}.
There are two approaches which can be taken to detect signatures of epistasis: one can look for the induced correlations in site patterns (sitewise measures), or one can look for evidence of substitutions co-localized to a branch (branchwise measures).
Our approaches include one sitewise and three branchwise measures, and we also apply the more general test statistic introduced by \cite{goldman1993statistical} as a general test of model adequacy.


\subsubsection*{Excess statistical coupling of site patterns}

In the absence of epistasis, pairs of sites will have dependence structure induced by the phylogeny.
With pair-wise epistasis, we expect excess coupling between epistatically paired sites.
We don't know which sites are paired, but we can calculate the mutual information for all pairs of sites.
The skew of this MI distribution is a measure of the effects of a subset of sites with pairwise epistasis.
\wdcomment{Notation here, especially $\ell$, is inconsistent with Andy's $S$.}
Denote the length $\ell$ alignment of $n$ taxa by the $n\times\ell$ character state matrix $D$, where $D_{ij}\in\mathcal{A}$.
The character states $\mathcal{A}$ are probably nucleotides.
Let $f_i(a)$ denote the relative frequency of character $a$ at site $i$, and $f_{ij}(a,b)$ denote the relative joint frequency of character $a$ at site $i$ and character $b$ at site $j$:
\[
f_i(a) = \frac{1}{n}\sum_{k=1}^{n}\boldsymbol{1}_{\{D_{ki}\}}(a),
\]
\[
f_{ij}(a,b) = \frac{1}{n}\sum_{k=1}^{n}\boldsymbol{1}_{\{D_{ki}\}}(a) \boldsymbol{1}_{\{D_{kj}\}}(b)
\]
The mutual information between sites $i$ and $j$ is then
\[
I_{ij} = \sum_{(a,b)\in\mathcal{A}^2}f_{ij}(a,b)\log\left(\frac{f_{ij}(a,b)}{f_i(a)f_j(b)}\right)
\]

Direct coupling analysis (\url{https://en.wikipedia.org/wiki/Direct_coupling_analysis}) would be a nice way to go too.
Although in our simulations there are only mated pairwise interactions (so mutual information should be sufficient and DCA is overkill), it should do just as well here and be more generalizable to more complicated epistasis.
We also score hipster points, and get to say impressive things like ``Ising model".

For DCA, Kristian recommends the code from Deb Marks' group: \url{https://github.com/debbiemarkslab/plmc} and \url{https://github.com/debbiemarkslab/EVcouplings}

In any case, some scalar like the skewness on either MI or the coupling terms in DCA should be sensitive to epistasis.
The number of couplings is $\frac{\ell(\ell-1)}{2}$.

\subsubsection*{Shared differences}
The next statistic is a statistic on comparisons of three taxa.
\wdcomment{The following sentence didn't do anything for me. Gesticulation and beer please.}
If pairwise epistasis exists in an alignment, where two taxa differ from a third, they should tend to have the differences more often at the same site than at different sites.
We can thus explicitly average over all three-taxon comparisons with taxa $i$, $j$, and $k$.
In this regime, we are looking for the proportion of sites (out of all sites at which either $i$ or $j$ differ from $k$) at which $i$ and $j$ have the same state.
We call this the Average Proportion of Shared Differences (APSD), and it can be calculated as,
\wdcomment{I started trying to simplify this notation, even though I don't yet understand it. But then I gave up. Definitely will require gesticulation and beer.}
\[
APSD = \frac{1}{\binom{n}{2}(n-3)} \sum_{i=1}^{n-1} \sum_{j=i+1}^n \sum_{k != i,j} \frac{D_S}{D}
\]
with
\[
D_S = \sum_{s=1}^S \mathbb{I}(A_{i,s} = A_{j,s} \neq A_{k,s})
\]
and
\[
D = \sum_{s=1}^S \mathbb{I}(A_{k,s} \neq A_{i,s},A_{j,s}).
\]
Where $A$ is the $n \times S$ alignment, $i$, $j$, and $k$ index the rows of the taxa being compared, and $\mathbb{I}$ is the indicator function.
For a given pair of taxa ($i$ and $j$), we count $D$ the number of sites in the alignment where either differs from the reference taxon ($k$).
We also count $D_S$ the number of these sites where $i$ and $j$ share a state, and divide by $D$ to calculate the proportion of shared differences.
We then average over all pairs of taxa and all reference taxa possible for each pair.

\subsubsection*{Shared polymorphisms}
However, the above calculation cubic in the number of taxa in the alignment, which while not prohibitively slow is certainly not fast.
One way to reduce the complexity is to avoid the sum over all reference taxa $t_j$.
How can this be done quickly?
If we restrict the summation to only occur over polymorphic sites (non-invariant sites) in the alignment, then we count the proportion of sites where $t_i$ and $t_j$ share a state and at least one differs from another taxon.
In this case, there is no explicitly defined third taxon for the comparison and we can simply average over pairwise comparisons.
We call this statistic the average proportion of shared polymorphic sites, or APSPS.
This can be calculated as,
\[
APSPS = \frac{1}{\binom{n}{2}} \sum_{i=1}^{n-1} \sum_{j=i+1}^n \frac{1}{P} \sum_{s=1}^P \mathbb{I}(A_{i,s} = A_{j,s})
\]
Where $A$ is the $n \times S$ alignment, $P$ is the number of polymorphic sites in $A$, $i$ and $j$ index the rows of the taxa being compared, and $\mathbb{I}$ is the indicator function.
For a given pair of taxa, we sum the number of polymorphic sites for which they share the same state and divide by the number of polymorphic sites to get a proportion, and we then take the average of this value.
While this is a crude approximation to marginalizing over the reference taxon, it is $O(n^2P)$ instead of $O(n^3S)$ ($P \leq S$).


\subsubsection*{Singleton mutations}
Our last posterior predictive approach to detecting epistasis requires a tree with branch lengths, which we will assume to be known for the purposes of describing the measure.
Given an edge $e$ in a phylogeny of length $l$ (measured in the expected number of substitutions per site) the number of changes along that branch is (approximately) distributed Poisson($e$).
Given two branches, $e_i$ and $e_j$, the total number of changes is (approximately) distributed Poisson($e_i + e_j$).
More importantly, we can compute the probability $\pr(i \mid c)$ that, given that there is a single change, it occurs on $e_i$: $\pr(i \mid c) = \frac{e_i}{e_i + e_j}$.
More generally, for a vector of branches $\boldsymbol{e}$, $\pr(i \mid c) = \frac{e_i}{\sum_j e_j}$.
%\amcommment{It seems possible that this probability statement is better behaved than the Poisson approximation given that we are making the same approximation in the numerator and the denominator and both edges are evolving under the same CTMC. If Will is looking for fun things to prove.}

% Thus, a vector $\boldsymbol{E}$ of observed changes on edges $\boldsymbol{e}$ should be distributed Multinomial($\frac{1}{\sum_i e_i} \boldsymbol{e}$).
% Any process that affects the CTMC that generates $\boldsymbol{E}$ will affect this expectation.



but for that tree should produce a value that has a known distribution and thus we can calculate a p-value against the hypothesis of a non-epistatic model.
As we in practice do not know the tree or its branch lengths, we calculate this statistic for each posterior sample and examine the posterior distribution on p-values.

On an edge $e$ of length $l$ in a phylogeny, the expected number of changes is $l$ at a single site.

Let $e_i$ be a pendant (terminal, has a leaf as a child) edge in the phlogeny with length $l_i$, with the vector of all pendant edges $\boldsymbol{e}$ and all lengths $\boldsymbol{l}$.



\section*{Results\label{sec:results}}

Do we need results?

\section*{Discussion\label{sec:discussion}}

We win, they lose.

\section*{Supplementary Material}



\section*{Acknowledgments}
We thank the physicists for being obnoxious enough to motivate us to do this project.

\bibliographystyle{plainnat}
\bibliography{refs}






\end{document}
